File: src/csc_gmres.f
!                     ********************
                      SUBROUTINE CSC_GMRES
!                     ********************
     & (MA,VB,VX,M,NITER,TOL,LUV,VD,PC,NT,RES)
!
!***********************************************************************
! CSC PACKAGE - SERGIO CASTIBLANCO
!***********************************************************************
!
!brief    1) SOLVES SYSTEM Ax=b WITH PRECONDITIONED GENERALIZED MINIMAL
!            RESIDUAL (GMRES) METHOD (RESTARTED) 
!
!history  Sergio Castiblanco
!+        17/02/2022
!+        Translation for original Matlab implementation
!
!+
! This function solves the system Ax=b with the Restarted Generalized
! Minimal Residual (Restarted GMRES)
!
! Entries:
!     MA : Matrix A in CSC storage
!     VB : Right hand side vector
!     VX : Final solution (must have the first guest for the solution)
!     M : Krylov space dimension
!     NITER : Max. number of iterations
!     TOL : tolerance for the stop through the norm of the residual
!     LUV  : values of LU decomposition matrix (for SSOR or ILU(0)) in
!             CSC_packed storage, if not SSOR or ILU(0), it must be have
!             arbitrary values.
!     VD  : DIAGONAL VALUES OF A, FOR (ABS)DIAGONAL PRECONDITIONINGS
!     PC  : preconditioning type
!       - 0 : No preconditioning
!       - 1 : Diagonal preconditioning
!       - 2 : Absolute diagonal preconditioning
!       - 3 : Symmetric SOR (Symmetric Gauss-Seidel -> SSOR, w=1)
!       - 4 : ILU(0)
!
!      Sergio A. Castiblanco B.
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| MA       |-->| MATRIX A IN CSC STORAGE                              |
!| VB       |-->| RIGTH HAND SIDE VECTOR                               |
!| VX       |<->| SOLUTION VECTOR, AND INTIAL GUESS                    |
!| M        |-->| KRYLOV SUBSPACE SIZE                                 |
!| NITER    |-->| MAXIMUM NUMBER OF ITERATIONS                         |
!| TOL      |-->| TOLERANCE FOR THE NORM OF RESIDUAL                   |
!| LUV      |-->| LU DECOMPOSITION VALUES (FOR SSOR OR ILU0)           |
!| VD       |-->| VECTOR WITH (ABS)DIAGONAL OF A (IF PC=1 OR PC=2)     |
!| PC       |-->| PRECONDITIONING OPTION                               |
!| NT       |<--| FINAL NUMBER OF OUTER ITERATIONS                     |
!| RES      |<--| FINAL NORM OF THE RESIDUAL                           |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE CSC_STORAGE, EX_CSC_GMRES => CSC_GMRES
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      TYPE(CSC_OBJ), INTENT(IN) :: MA
      INTEGER, INTENT(IN) :: M,PC,NITER
      INTEGER, INTENT(OUT) :: NT
      DOUBLE PRECISION, INTENT(IN) :: TOL
      DOUBLE PRECISION, INTENT(OUT) :: RES
      DOUBLE PRECISION, DIMENSION(MA%NR), INTENT(IN) :: VB
      DOUBLE PRECISION, DIMENSION(MA%NC), INTENT(IN) :: VD
      DOUBLE PRECISION, DIMENSION(MA%NZ), INTENT(IN) :: LUV
      DOUBLE PRECISION, DIMENSION(MA%NC), INTENT(INOUT) :: VX
!
!  IN SUBROUTINE VARIABLES
!
      DOUBLE PRECISION, DIMENSION(MA%NR+1,M+1) :: WH
      DOUBLE PRECISION, DIMENSION(M+1) :: PBET
      DOUBLE PRECISION, DIMENSION(M+1) :: VG
      DOUBLE PRECISION :: D,S,C,HIJ,HI1J
      DOUBLE PRECISION, DIMENSION(MA%NR) :: VZ,VQ,VR
      DOUBLE PRECISION :: BQTW
      INTEGER :: N,NB,T,I,J,K
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
! INIT
      WH = 0.0D0
      PBET = 0.0D0
      VG = 0.0D0
      N = MA%NC
      NB = SIZE(VB)
      IF(N.NE.NB)THEN
        WRITE(*,*) "ERROR GMRES, DIMENSSIONS DO NOT AGREE"
        STOP 0
      ENDIF
!-----------------------------------------------------------------------
! NO PRECONDITIONING
!-----------------------------------------------------------------------
      IF(PC==0)THEN
!
      DO T=1,NITER
!
        !RESIDUAL
        CALL CSC_MMATVEC(MA,VX,VR,N)
        VR = VB - VR
!
        !ARNOLDI METHOD WITH HOUSEHOLDER
        CALL CSC_ARNOLDIHOUSE(MA,VR,M,LUV,VD,PC,WH,PBET)
        !IF(T.LE.10) WRITE(*,*) "PBET",PBET
        !IF(T.LE.10) WRITE(*,*) "WH", Wh(1:(M+M),:)
!
        !ELIMINATION PROCESS FOR HESSENBERG MATRIX (GIVEN'S ROTATIONS)
        VG(1) = WH(1,1)
        DO I=1,M
          D = SQRT(WH(I,I+1)**2 + WH(I+1,I+1)**2)
          S = WH(I+1,I+1)/D
          C = WH(I,I+1)/D
          WH(I,I+1) = WH(I,I+1)*C + WH(I+1,I+1)*S
          WH(I+1,I+1) = 0.0D0
          DO J=I+1,M
            !UPDATING HESSENBERG MATRIX
            HIJ = WH(I,J+1)
            HI1J = WH(I+1,J+1)
            WH(I,J+1) = C*HIJ + S*HI1J
            WH(I+1,J+1) = -S*HIJ + C*HI1J
          ENDDO
          !UPDATING RHS VECTOR VG
          VG(I+1) = -S*VG(I)
          VG(I) = C*VG(I)
        ENDDO
!
        !BACKWARD SUBSTITUTION
        DO I=M,1,-1
          VG(I) = VG(I)/WH(I,I+1)
          VG(1:I-1) = VG(1:I-1) - WH(1:I-1,I+1)*VG(I)
        ENDDO
!
        !COMPUTING Z FOR X_M = X0 + Z
        VZ = 0.D0
        DO K = M,1,-1
          VZ(K) = VZ(K) + VG(K)
          VQ = VZ
          BQTW = PBET(K)*(DOT_PRODUCT(VQ(K:N),WH(K+1:N+1,K)))
          DO I=K,N
            VZ(I) = VQ(I) - WH(I+1,K)*BQTW
          ENDDO
        ENDDO
!
        !UPDATING THE SOLUTION
        VX = VX + VZ
!
        !TOLERANCE OVER RESIDUAL
        RES = ABS(VG(M+1))
        IF(RES.LT.TOL)THEN
          EXIT
        ENDIF
!
      ENDDO
      NT = T
!
!-----------------------------------------------------------------------
! DIAGONAL OR ABSOULTE-DIAGONAL PRECONDITIONING
!-----------------------------------------------------------------------
      ELSE IF((PC.EQ.1).OR.(PC.EQ.2))THEN
!
      DO T=1,NITER
!
        !RESIDUAL
        CALL CSC_MMATVEC(MA,VX,VR,N)
        VR = VB - VR
!
        !ARNOLDI METHOD WITH HOUSEHOLDER
        CALL CSC_ARNOLDIHOUSE(MA,VR,M,LUV,VD,PC,WH,PBET)
!
        !ELIMINATION PROCESS FOR HESSENBERG MATRIX (GIVEN'S ROTATIONS)
        VG(1) = WH(1,1)
        DO I=1,M
          D = SQRT(WH(I,I+1)**2 + WH(I+1,I+1)**2)
          S = WH(I+1,I+1)/D
          C = WH(I,I+1)/D
          WH(I,I+1) = WH(I,I+1)*C + WH(I+1,I+1)*S
          WH(I+1,I+1) = 0.0D0
          DO J=I+1,M
            !UPDATING HESSENBERG MATRIX
            HIJ = WH(I,J+1)
            HI1J = WH(I+1,J+1)
            WH(I,J+1) = C*HIJ + S*HI1J
            WH(I+1,J+1) = -S*HIJ + C*HI1J
          ENDDO
          !UPDATING RHS VECTOR VG
          VG(I+1) = -S*VG(I)
          VG(I) = C*VG(I)
        ENDDO
!
        !BACKWARD SUBSTITUTION
        DO I=M,1,-1
          VG(I) = VG(I)/WH(I,I+1)
          VG(1:I-1) = VG(1:I-1) - WH(1:I-1,I+1)*VG(I)
        ENDDO
!
        !COMPUTING Z FOR X_M = X0 + Z
        VZ = 0.D0
        DO K = M,1,-1
          VZ(K) = VZ(K) + VG(K)
          VQ = VZ
          BQTW = PBET(K)*(DOT_PRODUCT(VQ(K:N),WH(K+1:N+1,K)))
          DO I=K,N
            VZ(I) = VQ(I) - WH(I+1,K)*BQTW
          ENDDO
        ENDDO
!
        !RIGHT PRECONDITIONING
        VZ = VZ/VD
!
        !UPDATING THE SOLUTION
        VX = VX + VZ
!
        !TOLERANCE OVER RESIDUAL
        RES = ABS(VG(M+1))
        IF(RES.LT.TOL)THEN
          EXIT
        ENDIF
!
      ENDDO
      NT = T
!
!-----------------------------------------------------------------------
! SSOR OR ILU(0) PRECONDITIONING
!-----------------------------------------------------------------------
      ELSE IF((PC.EQ.3).OR.(PC.EQ.4))THEN
!
      DO T=1,NITER
!
        !RESIDUAL
        CALL CSC_MMATVEC(MA,VX,VR,N)
        VR = VB - VR
!
        !ARNOLDI METHOD WITH HOUSEHOLDER
        CALL CSC_ARNOLDIHOUSE(MA,VR,M,LUV,VD,PC,WH,PBET)
!
        !ELIMINATION PROCESS FOR HESSENBERG MATRIX (GIVEN'S ROTATIONS)
        VG(1) = WH(1,1)
        DO I=1,M
          D = SQRT(WH(I,I+1)**2 + WH(I+1,I+1)**2)
          S = WH(I+1,I+1)/D
          C = WH(I,I+1)/D
          WH(I,I+1) = WH(I,I+1)*C + WH(I+1,I+1)*S
          WH(I+1,I+1) = 0.0D0
          DO J=I+1,M
            !UPDATING HESSENBERG MATRIX
            HIJ = WH(I,J+1)
            HI1J = WH(I+1,J+1)
            WH(I,J+1) = C*HIJ + S*HI1J
            WH(I+1,J+1) = -S*HIJ + C*HI1J
          ENDDO
          !UPDATING RHS VECTOR VG
          VG(I+1) = -S*VG(I)
          VG(I) = C*VG(I)
        ENDDO
!
        !BACKWARD SUBSTITUTION
        DO I=M,1,-1
          VG(I) = VG(I)/WH(I,I+1)
          VG(1:I-1) = VG(1:I-1) - WH(1:I-1,I+1)*VG(I)
        ENDDO
!
        !COMPUTING Z FOR X_M = X0 + Z
        VZ = 0.D0
        DO K = M,1,-1
          VZ(K) = VZ(K) + VG(K)
          VQ = VZ
          BQTW = PBET(K)*(DOT_PRODUCT(VQ(K:N),WH(K+1:N+1,K)))
          DO I=K,N
            VZ(I) = VQ(I) - WH(I+1,K)*BQTW
          ENDDO
        ENDDO
!
        !RIGHT PRECONDITIONING
        CALL CSC_SOLPACKLU(MA%NR,MA%NC,MA%NZ,LUV,MA%R,MA%C,VZ,VZ)
!
        !UPDATING THE SOLUTION
        VX = VX + VZ
!
        !TOLERANCE OVER RESIDUAL
        RES = ABS(VG(M+1))
        IF(RES.LT.TOL)THEN
          EXIT
        ENDIF
!
      ENDDO
      NT = T
!
!-----------------------------------------------------------------------
      ENDIF
!-----------------------------------------------------------------------
      IF(NT.EQ.NITER)THEN
        WRITE(*,*) "WARNING: GMRES REACH MAX. ITERATIONS ", NT
        WRITE(*,*) "WARNING: RESIDUAL ", RES
      ENDIF
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE CSC_GMRES

File: matlab/csc_gmres.m
function [x,t,res] = csc_gmres(Av,Ar,Ac,b,x,m,niter,tol,LUv,pc)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% This function solves the system Ax=b with the Restarted Generalized
% Minimal Residual (Restarted GMRES)
%
% Entries:
%     Av,Ar,Ac : Matrix A in CSC storage
%     b : Right hand side vector
%     x : First guest for the solution
%     m : Krylov space dimension
%     niter : Max. number of iterations
%     tol : tolerance for the stop through the norm of the residual
%     LUv  : values of LU decomposition matrix (for SSOR or ILU(0)) in
%              CSC_packed storage, or values of diagonal of A if
%              (abs)Diagonal preconditioning.
%     pc  : preconditioning type
%       - 0 : No preconditioning
%       - 1 : Diagonal preconditioning
%       - 2 : Absolute diagonal preconditioning
%       - 3 : Symmetric SOR (Symmetric Gauss-Seidel -> SSOR, w=1)
%       - 4 : ILU(0)
%
%      Sergio A. Castiblanco B.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

n = length(Ac)-1;
nb = length(b);
if n~=nb
    t = -1;
    return
end

% NO PRECONDITIONING
if pc==0

for t = 1:niter

    %Residual
    r = b - csc_matvec(Av,Ar,Ac,x);
   
    %Arnoldi Method with Householder
    [WH,pbet] = csc_arnoldi_householder(Av,Ar,Ac,r,m,LUv,pc);
%     if t<=10
%         WH(1:2*m,:)
%         pbet
%     end
    
    %Elimination process for Hessemberg matrix H through Givens Rotations
    g = zeros(m+1,1);
    g(1) = WH(1,1);
    for i=1:m
        d = sqrt(WH(i,i+1)^2 + WH(i+1,i+1)^2);
        s = WH(i+1,i+1)/d;
        c = WH(i,i+1)/d;
        WH(i,i+1) = WH(i,i+1)*c + WH(i+1,i+1)*s;
        WH(i+1,i+1) = 0;
        for j=i+1:m
            %Updating Hessemberg Matrix
            hij = WH(i,j+1);
            hi1j = WH(i+1,j+1);
            WH(i,j+1) = c*hij + s*hi1j;
            WH(i+1,j+1) = -s*hij + c*hi1j;
        end
        %Updating RHS vector g
        g(i+1) = -s*g(i);
        g(i) = c*g(i);
    end
    
    %Backward sustitution %column major
    for i = m:-1:1
%         g(i) = ( g(i)-WH(i,i+2:m+1)*g(i+1:m) )/WH(i,i+1);
        g(i) = g(i)/WH(i,i+1);
        g(1:i-1) = g(1:i-1) - WH(1:i-1,i+1)*g(i); 
    end
    %here y = g(1:m)
    
    %Computing z for xm = x0+z
    z = zeros(n,1);
    for k = m:-1:1
        z(k) = z(k) + g(k);
        q = z;
        bqTw = pbet(k)*(q(k:n)'*WH(k+1:n+1,k));
        for i = k:n
            z(i) = q(i) - WH(i+1,k)*bqTw;
        end
    end
    
    %updating the solution
    x = x + z;
    
    %tolerance over residual
    res = abs(g(m+1));
    if res<tol
        break
    end
end

%DIAGONAL PRECONDITIONING OR ABSOLUTE DIAGONAL PRECONDITIONING
elseif pc==1 || pc==2

% D = csc_diaga(Av,Ar,Ac);
for t = 1:niter

    %Residual
    r = b - csc_matvec(Av,Ar,Ac,x);
    
    %Arnoldi Method with Householder
    [WH,pbet] = csc_arnoldi_householder(Av,Ar,Ac,r,m,LUv,pc);
    
    %Elimination process for Hessemberg matrix H through Givens Rotations
    g = zeros(m+1,1);
    g(1) = WH(1,1);
    for i=1:m
        d = sqrt(WH(i,i+1)^2 + WH(i+1,i+1)^2);
        s = WH(i+1,i+1)/d;
        c = WH(i,i+1)/d;
        WH(i,i+1) = WH(i,i+1)*c + WH(i+1,i+1)*s;
        WH(i+1,i+1) = 0;
        for j=i+1:m
            %Updating Hessemberg Matrix
            hij = WH(i,j+1);
            hi1j = WH(i+1,j+1);
            WH(i,j+1) = c*hij + s*hi1j;
            WH(i+1,j+1) = -s*hij + c*hi1j;
        end
        %Updating RHS vector g
        g(i+1) = -s*g(i);
        g(i) = c*g(i);
    end
    
    %Backward sustitution %column major
    for i = m:-1:1
%         g(i) = ( g(i)-WH(i,i+2:m+1)*g(i+1:m) )/WH(i,i+1);
        g(i) = g(i)/WH(i,i+1);
        g(1:i-1) = g(1:i-1) - WH(1:i-1,i+1)*g(i); 
    end
    %here y = g(1:m)
    
    %Computing z for xm = x0+z
    z = zeros(n,1);
    for k = m:-1:1
        z(k) = z(k) + g(k);
        q = z;
        bqTw = pbet(k)*(q(k:n)'*WH(k+1:n+1,k));
        for i = k:n
            z(i) = q(i) - WH(i+1,k)*bqTw;
        end
    end
    z = z./LUv;       %preconditioning change
    
    %updating the solution
    x = x + z;
    
    %tolerance over residual
    res = abs(g(m+1));
    if res<tol
        break
    end
end

%SSOR OR ILU(0) PRECONDITIONING
elseif pc==3 || pc==4

for t = 1:niter

    %Residual
    r = b - csc_matvec(Av,Ar,Ac,x);
    
    %Arnoldi Method with Householder
    [WH,pbet] = csc_arnoldi_householder(Av,Ar,Ac,r,m,LUv,pc);
    
    %Elimination process for Hessemberg matrix H through Givens Rotations
    g = zeros(m+1,1);
    g(1) = WH(1,1);
    for i=1:m
        d = sqrt(WH(i,i+1)^2 + WH(i+1,i+1)^2);
        s = WH(i+1,i+1)/d;
        c = WH(i,i+1)/d;
        WH(i,i+1) = WH(i,i+1)*c + WH(i+1,i+1)*s;
        WH(i+1,i+1) = 0;
        for j=i+1:m
            %Updating Hessemberg Matrix
            hij = WH(i,j+1);
            hi1j = WH(i+1,j+1);
            WH(i,j+1) = c*hij + s*hi1j;
            WH(i+1,j+1) = -s*hij + c*hi1j;
        end
        %Updating RHS vector g
        g(i+1) = -s*g(i);
        g(i) = c*g(i);
    end
    
    %Backward sustitution %column major
    for i = m:-1:1
%         g(i) = ( g(i)-WH(i,i+2:m+1)*g(i+1:m) )/WH(i,i+1);
        g(i) = g(i)/WH(i,i+1);
        g(1:i-1) = g(1:i-1) - WH(1:i-1,i+1)*g(i); 
    end
    %here y = g(1:m)
    
    %Computing z for xm = x0+z
    z = zeros(n,1);
    for k = m:-1:1
        z(k) = z(k) + g(k);
        q = z;
        bqTw = pbet(k)*(q(k:n)'*WH(k+1:n+1,k));
        for i = k:n
            z(i) = q(i) - WH(i+1,k)*bqTw;
        end
    end
    z = csc_solpacklu(LUv, Ar, Ac, z);       %preconditioning change
    
    %updating the solution
    x = x + z;
    
    %tolerance over residual
    res = abs(g(m+1));
    if res<tol
        break
    end
end

end

end